opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
+
+  /**
+   * Encodes the span into a binary representation of the span's identity and trace
+   * attributes.
+   */
+  byte[] toBinary(Span span);
```

Comment:
We chose `Map[String, String]` as the lowest common denominator, because the tracing implementation does not know how the actual RPC framework is going to serialize that map on the wire. If you have a better API that can be used in the `http_client.py` I included above, I would love to see it. I am not clear how `SpanFromMapExtractor` could be used, because it is implementation-specific, since it has knowledge of what keys can be found in the map, like `trace_id` and `span_id` (there's no requirement in OT as to what keys will be in the map once the span is converted to it). If you're saying that `SpanFromMapExtractor` is just an interface that RPC framework can somehow retrieve from the tracer, then I don't see how it's different from the tracer directly exposing `toMap` and `fromMap` methods that we have today (the only difference is composition, but I think you're objecting to the notion of map, not where the methods are defined).

It is to avoid these confusions that we published the [use cases](http://opentracing.io/use-cases/) on the site, which show concrete examples of end-user code using OT API. Can you try to rewrite the following part from the Tracing Client Calls use case using a different API?

```python
        # Goal: framework encodes the Span into HTTP headers of the outbound request
        h_ctx, h_attr = tracer.propagate_span_as_text(span)

        for key, value in h_ctx.iteritems():
            request.add_header(key, value)
        if h_attr:
            for key, value in h_attr.iteritems():
                request.add_header(key, value)
```